zookeeper-api/src/main/java/org/apache/helix/zookeeper/zkclient/util/ZkPathRecursiveWatcherTrie.java

```diff
@@ -0,0 +1,180 @@
+    Set<RecursivePersistListener> recursiveListeners;
+    // A list of recursive persist watcher on the current path
+
+    /**
+     * Create a trie node with parent as parameter.
+     *
+     * @param value the value stored in this node
+     */
+    private TrieNode(String value) {
+      this.value = value;
+      this.children =
+          new HashMap<>(4);  // We keep the same number of init children side as Zk serer
+      this.recursiveListeners = new HashSet<>(4);
+    }
+
+
+    /**
+     * The value stored in this node.
+     *
+     * @return the value stored in this node
+     */
+    public String getValue() {
+      return this.value;
+    }
+
+    /**
+     * Add a child to the existing node.
+     *
+     * @param childName the string name of the child
+     * @param node the node that is the child
+     */
+    void addChild(String childName, TrieNode node) {
+      this.children.putIfAbsent(childName, node);
+    }
+
+    /**
+     * Return the child of a node mapping to the input child name.
+     *
+     * @param childName the name of the child
+     * @return the child of a node
+     */
+    @VisibleForTesting
+    TrieNode getChild(String childName) {
+      return this.children.get(childName);
+    }
+
+    /**
+     * Get the list of children of this trienode.
+     *
+     * @return A collection containing the node's children
+     */
+    @VisibleForTesting
+    Collection<String> getChildren() {
+      return children.keySet();
+    }
+
+    /**
+     * Get the set of RecursivePersistWatcherListener
+     * Returns an empty set if no listener is registered on the path
+     * @return
+     */
+    @VisibleForTesting
+    Set<RecursivePersistListener> getRecursiveListeners() {
+      return recursiveListeners;
+    }
+
+    @Override
+    public String toString() {
+      return "TrieNode [name=" + value + ", children=" + children.keySet() + "]";
+    }
+  }
+
+  /**
+   * Construct a new PathTrie with a root node.
+   */
+  public ZkPathRecursiveWatcherTrie() {
+    this.rootNode = new TrieNode( "/");
+  }
+
+  /**
+   * Add a path to the path trie. All paths are relative to the root node.
+   *
+   * @param path the path to add RecursivePersistListener
+   * @param listener the RecursivePersistListener to be added
+   */
+  public void addRecursiveListener(final String path, RecursivePersistListener listener) {
+    Objects.requireNonNull(path, "Path cannot be null");
+
+    if (path.length() == 0) {
+      throw new IllegalArgumentException("Invalid path: " + path);
+    }
+    final String[] pathComponents = split(path);
+
+    synchronized(this) {
+      TrieNode parent = rootNode;
+      for (final String part : pathComponents) {
+        TrieNode child = parent.getChild(part);
+        if (child == null) {
+          child = new TrieNode(part);
+          parent.addChild(part, child);
+        }
+        parent = child;
```

Comment:
This code can be simplified with returning entire map as:

parent.getChildren().computeIfAbsent(part, tn -> new HashMap<>()).add(part, child);
parent = child;